x/qgb/abci.go

```diff
@@ -113,3 +122,108 @@ func handleValsetRequest(ctx sdk.Context, k keeper.Keeper) {
+	case types.ValsetRequestType:
+		vs, ok := lastAttestationInStore.(*types.Valset)
+		if !ok {
+			panic("couldn't cast valset type")
+		}
+		lastNonceHeight = vs.Height
+	}
+	lastUnbondingHeight := k.GetLastUnBondingBlockHeight(ctx)
+	if lastNonceHeight == lastUnbondingHeight {
+		// we don't need to do anything, we want to keep attestations up to the last unbonding height
+		return
+	}
+	if lastNonceHeight > lastUnbondingHeight {
+		// checking if it's the initial case following the startup of the chain
+		if lastNonceHeight == 1 && lastUnbondingHeight == 0 {
+			return
+		}
+		// we should never hit this case, since we will keep the attestations up to the last unbonding height
+		panic("missing attestations up to the unbonding height")
+	}
+	// now we have attestations before the unbonding height, we should check whether we need to prune them
+	// or not yet
+	latestAttestationNonce := k.GetLatestAttestationNonce(ctx)
+	if latestAttestationNonce-lastAvailableNonce+1 <= AttestationPruningThreshold {
+		// we don't need to prune as we still have room to store more attestations
+		return
+	}
+	// now we want to prune attestations as we have the following conditions:
+	// - we have attestations up to the last unbonding height
+	// - the total number of attestations, including the ones before the unbonding height, are greater
+	// than the AttestationPruningThreshold
+	err = pruneAttestations(ctx, k, lastAvailableNonce, latestAttestationNonce)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// pruneAttestations prunes attestations to keep attestations up to the last unbonding height
+// and the total number of attestations lower than the AttestationPruningThreshold.
+func pruneAttestations(
+	ctx sdk.Context,
+	k keeper.Keeper,
+	lastAvailableAttestationNonce uint64,
+	latestAttestationNonce uint64,
+) error {
+	ctx.Logger().Debug("pruning attestations from qgb store")
+	if !k.CheckLastUnbondingNonce(ctx) {
+		// We should never hit this case in the happy path because at this level, we're sure that
+		// all store values are initialized.
+		// However, we might hit it in an upgrade if we don't handle the store values carefully.
+		// So, it's good to have this check in place.
+		return fmt.Errorf("last unbonding nonce not initialized in store")
+	}
+	lastUnbondingNonce := k.GetLastUnbondingNonce(ctx)
+	count := 0
+	newLastAvailableNonce := lastAvailableAttestationNonce
+	for i := lastAvailableAttestationNonce; i < lastUnbondingNonce; i++ {
+		if newLastAvailableNonce == lastUnbondingNonce ||
+			latestAttestationNonce-i+1 == AttestationPruningThreshold {
+			// we either reached the last unbonding height, and we need to keep all those attestations
+			// or, we reached the minimum number of attestations we want to keep in store
+			newLastAvailableNonce = i
+			break
+		}
+		k.DeleteAttestation(ctx, i)
+		count++
+	}
+	// persist the new last pruned attestation nonce
+	k.SetLastPrunedAttestationNonce(ctx, newLastAvailableNonce-1)
```

Comment:
why subtract one here if we add it back in line 134? (see other comment there as well)

Reply:
This one is to set the value of the last pruned nonce which is: `last pruned nonce == last available nonce - 1`.

The other is to calculate the number of attestations that will be in store